### PR TITLE
Update Azure libs - part 2

### PR DESCRIPTION
**Pull request by FabioRosado**
This is my first stab at updating Azure dependencies following your awesome work @nuwang I have a bunch of tests still failing which I need to look into and I would like to do some manual testing just in case as well since a lot changed in the Azure API.

I'll be adding some comments/questions but feel free to skip this until I do more work on this PR :smile: 

I've also added some redundant changes to make typing happy, not sure if you care about it and I'm happy to revert these changes :smile:

### CODE REVIEW

File: cloudbridge/providers/azure/azure_client.py [hunk 1]
```diff
@@ -1,28 +1,27 @@
 import datetime
 import logging
-from io import BytesIO
+
+import tenacity
+from cloudbridge.interfaces.exceptions import (DuplicateResourceException,
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:5]: This was my automatic formatter, should I change it back?

File: cloudbridge/providers/azure/azure_client.py [hunk 2]
```diff
@@ -181,13 +180,13 @@ def __init__(self, config):
         log.debug("azure subscription : %s", self.subscription_id)
 
     @property
-    @tenacity.retry(stop=tenacity.stop_after_attempt(5), reraise=True)
+    @tenacity.retry(stop=tenacity.stop.stop_after_attempt(5), reraise=True)
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:183]: pylance was complaining and this seems to make it happy :shrug:  I'm happy to revert the change

File: cloudbridge/providers/azure/azure_client.py [hunk 3]
```diff
@@ -310,8 +311,9 @@ def _get_or_create_storage_account(self):
             try:
                 self._storage_account = \
                     self.get_storage_account(self.storage_account)
-            except CloudError as cloud_error:
-                if cloud_error.error.error == "ResourceNotFound":
+            except HttpResponseError as cloud_error:
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:314]: I suspect we might need to change all of the `CloudError` to `HttpResponseError` since this is what Azure seems to raise now

File: cloudbridge/providers/azure/azure_client.py [hunk 4]
```diff
@@ -367,47 +368,47 @@ def list_vm_firewall(self):
...
-            create_or_update(self.resource_group, name,
-                             parameters).result()
+            begin_create_or_update(self.resource_group, name,
+                                   parameters).result()
 
     def update_vm_firewall_tags(self, fw_id, tags):
         url_params = azure_helpers.parse_url(VM_FIREWALL_RESOURCE_ID,
                                              fw_id)
-        name = url_params.get(VM_FIREWALL_NAME)
+        name = url_params.get(VM_FIREWALL_NAME, "")
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:377]: This is one of the many "make typing happy" kind of change, let me know if you want me to revert this

File: cloudbridge/providers/azure/azure_client.py [hunk 5]
```diff
@@ -417,66 +418,73 @@ def list_containers(self, prefix=None, limit=None, marker=None):
...
 
     def get_blob(self, container_name, blob_name):
-        return self.blob_service.get_blob_properties(container_name, blob_name)
+        blob_client = self.blob_client(container_name, blob_name)
+        return blob_client.get_blob_properties(container_name, blob_name)
 
-    def create_blob_from_text(self, container_name, blob_name, text):
-        self.blob_service.create_blob_from_text(container_name,
-                                                blob_name, text)
+    def create_blob_from_text(self, container_name, blob_name, text, length=None):
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:449]: Azure mentions that if you pass the `length` keyword argument, it can improve performance, so I added this here, what do you think?

File: cloudbridge/providers/azure/azure_client.py [hunk 6]
```diff
@@ -417,66 +418,73 @@ def list_containers(self, prefix=None, limit=None, marker=None):
...
+        self.upload_blob(container_name, blob_name, text, length)
 
-    def create_blob_from_file(self, container_name, blob_name, file_path):
-        self.blob_service.create_blob_from_path(container_name,
-                                                blob_name, file_path)
+    def create_blob_from_file(self, container_name, blob_name, file_path, length=None):
+        with open(file_path, 'rb') as data:
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:453]: The new API doesn't seem to implement a `create_blob_from_path` but we can use `blob_upload` to upload data. 

I'm worried about the lack of checking, should we handle cases where the path isn't there? Or the assumption that the path exists is okay?

File: cloudbridge/providers/azure/azure_client.py [hunk 7]
```diff
@@ -417,66 +418,73 @@ def list_containers(self, prefix=None, limit=None, marker=None):
...
+            self.upload_blob(container_name, blob_name, data, length)
 
-    def delete_blob(self, container_name, blob_name):
-        self.blob_service.delete_blob(container_name, blob_name)
+    def delete_blob(self, container_name, blob_name, delete_snapshots="include"):
+        blob_client = self.blob_client(container_name, blob_name)
+        blob_client.delete_blob(delete_snapshots)
 
     def get_blob_url(self, container_name, blob_name, expiry_time):
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:460]: This is one of those changes that might work, but I haven't tested this yet.

File: cloudbridge/providers/azure/azure_client.py [hunk 8]
```diff
@@ -417,66 +418,73 @@ def list_containers(self, prefix=None, limit=None, marker=None):
...
+        return url
 
     def get_blob_content(self, container_name, blob_name):
-        out_stream = BytesIO()
-        self.blob_service.get_blob_to_stream(container_name,
-                                             blob_name, out_stream)
-        return out_stream
+        blob_client = self.blob_client(container_name, blob_name)
+        out_stream = blob_client.download_blob()
+        return out_stream.readall()
```
Comment (FabioRosado) [inline cloudbridge/providers/azure/azure_client.py:477]: This is a blocking call, should we returning the [StorageStreamDownloader](https://docs.microsoft.com/en-us/python/api/azure-storage-blob/azure.storage.blob.storagestreamdownloader?view=azure-python) and let users choose how to download the blob instead?

File: cloudbridge/providers/azure/azure_client.py (hunk 1 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:5]: This if fine if flake8 is fine with it. Please try running `tox -e lint`.

File: cloudbridge/providers/azure/azure_client.py (hunk 2 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:183]: May have been a change in the latest version of tenacity? Fine if it's happy.

File: cloudbridge/providers/azure/azure_client.py (hunk 3 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:314]: Makes sense. Maybe we can directly catch `ResourceNotFoundError`?

File: cloudbridge/providers/azure/azure_client.py (hunk 4 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:377]: I think it's fine, thanks for fixing it.

File: cloudbridge/providers/azure/azure_client.py (hunk 5 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:449]: Can we make this len(text) and compute it? The cloudbridge interface doesn't have a length parameter, so it seems like an implementation specific thing.
```
self.upload_blob(container_name, blob_name, text, length=len(text))
```

File: cloudbridge/providers/azure/azure_client.py (hunk 6 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:453]: My intuition is that the assumption is ok, because if it's not, the error message will be reasonable. This gets autowrapped in a CloudBridgeBaseException anyway, so it can still be caught in a generic way, but the error message should be pretty easily comprehensible?

File: cloudbridge/providers/azure/azure_client.py (hunk 8 above)
Comment (nuwang) [reply cloudbridge/providers/azure/azure_client.py:477]: Is the blob_contents method being called anymore? The iter_contents() method is the "non-blocking" version, and it's returning chunks now: https://github.com/CloudVE/cloudbridge/pull/277/files/eafa0811e41b07c8c82b566a9b889c4918f1966c#diff-6c6fa7ef69cdd88282c8567237b70cdb7501937a130df4ce156014b746649419R214

**Review by nuwang — COMMENTED** [review]
@FabioRosado Thanks for making these changes, they are all steps in the right direction. Would you mind reopening this PR against branch `update_azure_09_21`. That way, we can keep merging your changes and running tests. External PRs don't get access to cloud credentials, but branches do, so we can then get the tests to run.

**Comment by FabioRosado** [comment]
Thank you so much for the review Nuwan, I'll address your comments and do as you requested to opening this PR against that branch.

I'm going to pick this work on the weekend, hopefully, we can get some good progress done :smile:
